02-express-tutorial/app.js

```diff
@@ -1 +1,63 @@
-console.log('Express Tutorial')
+const { products } = require("./data");
+const express = require('express');
+
+console.log('Express Tutorial');
+
+const app = express();
+
+app.use(express.static('./public'));
+
+
+app.get('/api/v1/test', (req, res) => {
+    res.json(products);
+});
+
+
+app.get('/api/v1/products/:productID', (req, res) => {
+    const idToFind = parseInt(req.params.productID);
+    const product = products.find(p => p.id === idToFind);
+ 
+    if (!product) {
+        return res.status(404).json({ message: 'Product not found' });
+    }
+
+    res.json(product);
+});
+
+
+app.get('/api/v1/query', (req, res) => {
+    const { search, searchRegex, limit, maxPrice } = req.query;
+    let filteredProducts = [...products];
+
+    if (search) {
+        const searchTerm = search.toLowerCase();
+        filteredProducts = filteredProducts.filter(p => p.name.toLowerCase().startsWith(searchTerm));
+    }
+
+    if (searchRegex) {
+        const regex = new RegExp(searchRegex, 'i'); 
+        filteredProducts = filteredProducts.filter(p => regex.test(p.name));
+    }
+
+    if (maxPrice) {
+        const maxPriceNumber = parseFloat(maxPrice);
+        filteredProducts = filteredProducts.filter(p => p.price < maxPriceNumber);
+    }
+
+    if (limit) {
+        const limitNumber = parseInt(limit);
+        filteredProducts = filteredProducts.slice(0, limitNumber);
+    }
+
+    res.json(filteredProducts);
+});
+
+
+app.use((req, res, next) => {
```

Comment:
You are not using the app.all statement to handle routes that are not found. While this still technically works, there are certain methods and scenarios for using this vs all that you wouldn't be able to access since you are using this approach. But either way, the instructions say to use it.